adding_tags/add_user_tags.rb

```diff
@@ -0,0 +1,165 @@
+require 'optparse'
+require 'logger'
+
+class PagerDutyAgent
+  attr_reader :token
+  attr_reader :connection
+
+  def initialize(token)
+    @token = token
+    @connection = Faraday.new(:url => "https://api.pagerduty.com",
+                              :ssl => {:verify => true}) do |c|
+      c.request  :url_encoded
+      c.adapter  :net_http
+    end
+    #making an initial call to /users to extract subdomain corresponding to the API key
+    initial = connection.get("/users", query = {}, { 'Authorization' => "Token token=#{token}",
+          'Accept' => 'application/vnd.pagerduty+json;version=2'})
+    users = JSON.parse(initial.body)['users']
+    subdomain = users[0]['html_url'][/https:\/\/(.*)\.pagerduty\.com.*/, 1]
+    puts("Adding user tags for #{subdomain}, proceed? (y/n)")
+    #evaluating agent input
+    decision = ""
+    while decision.chomp != "y"
+      decision = gets
+      if decision.chomp == "n"
+        abort "Discontinuing user import"
+      end
+    end
+      #logger defined as a global variable accessible to all classes in the script
+      $log = Logger.new("import_errors_for_#{Date.today}.log")
+  end
+
+  def get(path, query = {})
+    response = connection.get(path, query,
+        { 'Authorization' => "Token token=#{token}",
+          'Accept' => 'application/vnd.pagerduty+json;version=2'})
+    if !response.success?
+      $log.error("Status: #{response.status}
+        Response: #{response.body}---Query: #{query}")
+      raise "Error: #{response.body}"
+    end
+    JSON.parse(response.body)
+  end
+
+  def post(path, body = {}, extra_headers = nil)
+    body_json = JSON.generate(body)
+    if @extra_headers
+      headers = { 'Authorization' => "Token token=#{token}",
+        'Content-Type' => 'application/json',
+        'Accept' => 'application/vnd.pagerduty+json;version=2'}
+    else
+      headers = { 'Authorization' => "Token token=#{token}",
+        'Content-Type' => 'application/json',
+        'Accept' => 'application/vnd.pagerduty+json;version=2'}
+    end
+    response = connection.post(path, body_json, headers)
+
+    if !response.success?
+      $log.error("Status: #{response.status}--Response: #{response.body}---Payload: #{body}")
+      raise "Error: #{response.body}"
+    end
+
+    return JSON.parse(response.body)
+  end
+
+  def find_user_by_email(email)
+    get("/users", :query => email)
+  end
+
+  def add_tags(user_id, tag)
+      tags = tag.map do |t|
+        {
+          :type => "tag",
+          :label => t
+        }
+      end
+
+    request_body = {
+      :add => tags,
+      :remove => [
+        {
+          :type => nil,
+          :id => nil
+        }
+      ]
+    }
+    post("/users/#{user_id}/change_tags", request_body)
+  end
+
+end
+
+class CSVImporter
+
+  Record = Struct.new(:email, :tags)
+
+  attr_reader :agent
+  attr_reader :csv_file
+
+  def initialize(agent, csv_file)
+    @agent = agent
+    @csv_file = csv_file
+  end
+
+  def import_user(record)
```

Comment:
If this is planned to be a standalone script for adding user tags, maybe we should change the "import_user" references to "add_user_tags"?

Reply:
I have now renamed the method to match the action for this script